docs/hashing.rst

```diff
@@ -0,0 +1,46 @@
+Hashing
+=======
+
+.. warning::
+
+   The overarching theme is to never set the ``@attr.s(hash=X)`` parameter yourself.
+   Leave it at ``None`` which means that ``attrs`` will do the right thing for you, depending on the other parameters.
+
+   Setting ``hash`` yourself can have unexpected consequences so we recommend to tinker with it only if you know exactly what you're doing.
+
+Under certain circumstances, it's necessary for objects to be *hashable*.
+For example if you want to put them into a :class:`set` or if you want to use them as keys in a :class:`dict`.
+
+The *hash* of an object is an integer that represents the contents of an object.
+It can be obtained by calling :func:`hash` on an object and is implemented by writing a ``__hash__`` method for your class.
+
+``attrs`` will happily write a ``__hash__`` method you [#fn1]_, however it will *not* do so by default.
+Because according to the definition_ from the official Python docs, the returned hash has to fullfill two constraints:
+
+#. The hash of an object *must not* change.
+   This is the reason why mutable structures like lists, dictionaries, or sets aren't hashable while immutable ones like tuples or frozensets are.
```

Comment:
The way this is ordered seems confusing for those who don't already know about hashes... if the first thing you say about a hash is that it can't change, then it's not clear why it shouldn't be `def __hash__(self): return 0` or what mutability has to do with it.

Maybe: (1) If objects are `==`, then their hashes MUST be the same. (2) If objects are not `==`, then their hashes SHOULD be different – technically this isn't a requirement, but it has to be true most of the time or sets and dicts will become horribly inefficient. (3) The hash MUST NOT change.

So (2) means that if you mutate an object in a way that changes `==` comparisons, then the hash has to change, but (3) says it can't change => that's why only immutable objects are hashable.